services/smtp.py

```diff
@@ -0,0 +1,87 @@
+
+    @classmethod
+    def active(cls):
+        return cls.connection is not None
+
+    def _load_config(self):
+        if get_config("services", "smtp", default={}) == {}:
+            return False
+        self.host = get_config("services", "smtp", "host", default="mailhog")
+        self.port = get_config("services", "smtp", "port", default=1025)
+        self.username = get_config("services", "smtp", "username", default=None)
+        self.password = get_config("services", "smtp", "password", default=None)
+        self.ssl_context = ssl.create_default_context()
+        return True
+
+    def __init__(self):
+        if not self._load_config():
+            log.warning("Unable to load SMTP config")
+            return
+        if SMTPService.connection is None:
+            SMTPService.connection = smtplib.SMTP(
+                host=self.host,
+                port=self.port,
+            )
+
+            # only necessary if SMTP server supports TLS and authentication,
+            # for example mailhog does not need these two steps
+            try:
+                SMTPService.connection.starttls(context=self.ssl_context)
+            except smtplib.SMTPNotSupportedError:
+                log.warning(
+                    "Server does not support TLS, continuing initialization of SMTP connection",
+                    extra=dict(
+                        host=self.host,
+                        port=self.port,
+                        username=self.username,
+                        password=self.password,
+                    ),
+                )
+
+            if self.username and self.password:
+                try:
+                    SMTPService.connection.login(self.username, self.password)
+                except smtplib.SMTPNotSupportedError:
+                    log.warning(
+                        "Server does not support auth, continuing initialization of SMTP connection",
+                        extra=dict(
+                            host=self.host,
+                            port=self.port,
+                            username=self.username,
+                            password=self.password,
+                        ),
+                    )
+
+    def send(self, email: Email):
+        err_msg = None
+        if not SMTPService.connection:
+            return "Connection was not initialized"
+        try:
+            errs = SMTPService.connection.send_message(
+                email.message,
+            )
+            if len(errs) != 0:
+                err_msg = " ".join(
+                    list(map(lambda err_tuple: f"{err_tuple[0]} {err_tuple[1]}", errs))
+                )
+        except smtplib.SMTPRecipientsRefused:
+            err_msg = "All recipients were refused"
+        except smtplib.SMTPSenderRefused:
+            err_msg = "Sender was refused"
+        except smtplib.SMTPDataError:
+            err_msg = "The SMTP server did not accept the data"
+
+        return err_msg
```

Comment:
should we just log this error message now instead of making the caller do it for us?

i think i'd prefer you create a new `SMTPServiceSendFailure` exception and raise it on error. it's more clear/aggressive than returning `None` on success and a value on failure, which is the opposite of how functions usually behave. but i think python's type system might be too weak to make me totally happy

_(**nothing actionable is below this point**, just explaining why i don't really like any options here)_

when i write error-handling code i keep two questions in mind:
- should the caller care about the outcome of this function?
- if so, how can i force the caller to deal with both success and failure?

on the first point, `SMTPService` doesn't have its own retry logic, so retrying is something the caller is responsible for. the caller should care if this function passes or fails. i think this makes sense; the caller is probably a celery task which has a ready-made `retry()` function but inside this function we'd need custom logic

but for the second point, python leaves me sad. compare to other languages:
- java makes you declare every type of exception your function can throw. so you either *must* place your call to `foo()` in a `try`/`catch` block that handles every exception `foo()` can throw, or you *must* declare your function to throw all the exceptions that `foo()` can throw (and punt the problem up a stack frame). either way, the compiler forces the caller to make a decision about how to handle the error
- C++ and Rust provide ways to require that you handle the return value after you call `foo()`. you can return `Result<Val, Err>` types annotated in this way and the compiler will make the caller explicitly decide how to handle them:
```
fn foo() -> Result<T, E>;

fn bar1() {
    foo(); // fails to build because return value is not assigned to anything
}

fn bar2() {
    let result = foo(); // fails because the result variable is never actually used
}

fn bar3() {
    let result = foo();
    if result.is_err() { // passes because we're actually using the result variable
        schedule_retry();
    }
}

fn bar4() {
    let _ = foo(); // passes because _ is a magic "i am intentionally ignoring this" identifier
}
```

python can make you handle neither return values nor exceptions. an uncaught exception will make the caller handle the error after they ship a bug into production, but python can't save them from the bug in the first place like other languages can